---
title: 정보은닉과 dict
---

## 속성 감추기

```python
class Person:
    def __init__(self,n,a):
        self.name = n
        self.age = a
    def __str__(self):
        return '{0}: {1}'.format(self.name, self.age)

def main():
    p = Person('Park',23)
    p.age-=1
    print(p)
```

위 코드의 문제점은 ? -> age에 빼기 연산을 진행했다는것!

사람의 나이는 거꾸로 먹을 수 없다.

해당 논리적 오류가 발생한 근본적인 이유가 무엇일까..

**객체 외부에서 객체 내에 있는 변수(속성)에 직접 접근하도록 내버려두었다.**

1차적으로, 나이 속성에 대해 더하기 연산을 진행하고 싶으면 add메소드를 생성한다.

이후에도 혹시 모를 나이 속성에 대한 직접 접근을 원천 봉쇄하기 위해 조치를 취하는 것

**객체의 인스턴스 변수 앞에 언더바를 하나 추가하면 그 변수는 감춰진다.**

```python
class Person:
    def __init__(self,n,a):
        self._name = n
        self._age = a
    def add_age(self,a):
        if a<0:
            print("잘못된 나이 정보 전달")
        else:
            self._age += a
```

인스턴스 변수 앞에 언더 바가 **하나만 있으면** 객체.\_인스턴스 형태의 코드로 인스턴스 변수를 직접 호출해도 오류는 발생하지 않는다. but -> 프로그래머간의 약속 형태로 언더바 하나가 추가된 인스턴스 변수에는 접근하지 말아야한다.

인스턴스 변수 앞에 언더 바가 **두개 있으면** 객체.\_\_인스턴스 형태의 코드로 인스턴스 변수를 직접 호출하면 오류가 발생한다!!

## **dict**

객체 내에는 해당 객체의 변수 정보를 (속성 정보를) 담고 있는 딕셔너리가 하나 존재한다.

```python
class Person:
    def __init__(self,n,a):
        self._name = n
        self._age = a


def main():
    p = Person('park', 23)
    print(p.__dict__)

```

위의 출력 결과는

```python
>>> {'_name': 'park', '_age': 23}
```

과 같이 나타난다.

이를 통해 두 가지 사실을 알 수 있는데,

1. 객체 내에 변수 \_name이 있는데, 그 안에는 park이 담겨 있다.

2. 객체 내에 변수 \_age가 있는데, 그 안에는 23이 담겨있다.

객체 내의 dict 스페셜 메소드는 클래스 정의 이후 **변수가 추가되거나 값이 수정되면 그대로 다시 반영한다**

또한 dict를 통해 **객체 내 값도 수정할 수 있다.**

```python
class Simple:
    def __init__(self,n,s):
        self._n = n
        self._s = s
    def __str__(self):
        return '{0}: {1}'.format(self._n, self._s)

def main():
    sp = Simple(10,'my')
    sp.__dict__['_n'] += 10
    sp.__dict__['_s'] = 'your'
```

만약 인스턴스 변수에 언더바를 두개 붙인다면?

**\_\_AttrName** => **\_ClassName\_\_AttrName** 형식으로 바뀌게 된다.

외부에서 접근시 오류가 발생했던 이유가 바로 언더바를 두개 붙일 시 dict접근에 대한 속성 이름을 파이썬에서 임의로 변경시켰기 때문이다

따라서 바뀐 패턴에 따라 접근하면 파이썬은 이를 막지 못함..!

**오류 발생도 AttributeError가 발생함.**
